### PR TITLE
Enable gocbv2 by default

### PR DESCRIPTION
**Pull request by adamcfraser**
Enables gocb v2 by default for everything except DCP.  Testing the full suite in this mode caught a number of bugs in the collection implementation, particularly around standardizing error types.  Beyond bug fixes, created a new CouchbaseStore interface for functionality specific to a backing Couchbase Server, and not already on the bucket interface.

Includes a transcoder for gocb v2 that is backward compatible with the gocb v1 transcoder.  (Attempted to pull that out into a separate PR, but it had too many interdependencies with the rest of the fixes and refactoring in this PR).

Dependencies:
- [x] https://github.com/couchbase/sg-bucket/pull/65

Integration tests:
GSI: http://uberjenkins.sc.couchbase.com:8080/view/Build/job/sync-gateway-integration/957/
Views: http://uberjenkins.sc.couchbase.com:8080/view/Build/job/sync-gateway-integration/958/
Views, no xattrs: http://uberjenkins.sc.couchbase.com:8080/view/Build/job/sync-gateway-integration/959/

### CODE REVIEW

File: manifest/default.xml
Comment (bbrks) [inline manifest/default.xml]: Update for merged PR

```suggestion
  <project name="sg-bucket" path="godeps/src/github.com/couchbase/sg-bucket" remote="couchbase" revision="08f9bc2aa1cb5cb87b9a9bc0f803ddefbc73c5f8"/>
```

**Review by bbrks — CHANGES_REQUESTED** [review]
LGTM. Can bump manifest, and couple of non-code nitpicks then should be good to merge if there are no more changes for this branch.

**Comment by adamcfraser** [comment]
Addressed PR changes but also made a couple of non-trivial changes (GetIndexes(), socket closes).  @bbrks Can you give this one more pass before I merge?

**Review by bbrks — APPROVED** [review]
LGTM

**Pull request merged**
